### PR TITLE
Paper: PyDDA: A New Pythonic Multiple Doppler Retrieval Package

### PR DESCRIPTION
**Pull request by rcjackson**
This paper shows a new Python package for retrieving winds from radar networks called PyDDA.

### CODE REVIEW

**Comment by rcjackson** [comment]
Not sure why it doesn't build. The build completes successfully on my machine.

**Comment by deniederhut** [comment]
GitHub has a backend outage earlier which may have been the issue. Looks fine now!

File: papers/robert_jackson/paper.rst [hunk 1]
```diff
@@ -0,0 +1,617 @@
...
+
+:bibliography: mybib
+
+------------------------------------------------
+PyDDA: A new Pythonic Wind Retrieval Package
+------------------------------------------------
+
+.. class:: abstract
+
+   PyDDA is a new community framework aimed at wind retrievals that depends
```
Comment (ocefpaf) [inline papers/robert_jackson/paper.rst:30]: ~~More of a question than a review comment but it is unclear to me why a specialized package instead of pushing for metadata standard and using a "generic" package to fetch the data. I'm Thinking about OPeNDAP/ERDDAP and CF-1.7. (Bare in mind I'm not a meteorologist nor I deal with wind data on a day-to-day basis.)~~

Answering my own question as I read more below. The word "retrievals" threw me off. This is calculating winds from data. (Quite interesting BTW!)

File: papers/robert_jackson/paper.rst [hunk 2]
```diff
@@ -0,0 +1,617 @@
...
+:cite:`Langetal2017`. Since CEDRIC uses a strong constraint
+from mass continuity equation to retrieve winds, the addition of constraints
+from other data sources is not possible with CEDRIC. Also, while CEDRIC was
+revolutionary for its time, it is difficult to use as it uses a scripting
+language as inputs for the retrieval. While MultiDop is based off of the
+more customizable 3D variational technique, it is fixed to 2 or 3 radars and
+is not scalable. In addition, the addition of constraints from other 3D
+fields such as models is not supported. Finally, Multidop is a wrapper
+around a program written in C which introduces issues related to packaging
+in frameworks such as anaconda and scalability due to the non-threadsafe
```
Comment (ocefpaf) [inline papers/robert_jackson/paper.rst]: Should it be non-thread-safe? Ugly but I don't think the word threadsafe exists.

File: papers/robert_jackson/paper.rst [hunk 3]
```diff
@@ -0,0 +1,617 @@
...
+
+
+.. class:: keywords
+
+   wind, retrieval, hurricane, tornado, radar
+
+Introduction
+------------
+
+Three dimensional wind retrievals are important for examining the dynamics
```
Comment (ocefpaf) [inline papers/robert_jackson/paper.rst:48]: I have to say that, after reading just the abstract of the paper mentioned here, the authors did an amazing job on avoiding/explain jargon!

File: papers/robert_jackson/paper.rst (hunk 2 above)
Comment (ocefpaf) [inline papers/robert_jackson/paper.rst]: A minor nitpick but anaconda has compilers and makes packaging C code less painful. I would just say "written in C which introduces issues related to packaging."

File: papers/robert_jackson/paper.rst [hunk 4]
```diff
@@ -0,0 +1,617 @@
...
+pip or anaconda. Given that this alone is a major hurdle to using currently
+existing retrieval software, this makes it easier for those who are not
+radar scientists to be able to use the software. Unlike currently existing software,
+a suite of unit tests are built into PyDDA that are executed whenever a user
+make a contribution to PyDDA, ensuring that the package will function for the
+user. With regards to ease of use, PyDDA can retrieve winds from multiple radars
+combined with data from model reanalyses with just a few lines of code. In addition,
+PyDDA is built upon the Python ARM Radar Toolkit (Py-ART) :cite:`HelmusandCollis2016`.
+Since Py-ART is already used by hundreds of users in the radar meteorology community, these
+users would be able to learn how to use PyDDA easily. In addition, the open source nature
```
Comment (ocefpaf) [inline papers/robert_jackson/paper.rst]: Suggestion: two "In addition" in a row.  Change one to something like "moreover."

File: papers/robert_jackson/paper.rst [hunk 5]
```diff
@@ -0,0 +1,617 @@
...
+        if(anel == 1):
+            rho = np.exp(-z/10000.0)
+            drho_dz = np.gradient(rho, dz, axis=0)
+            anel = w/rho*drho_dz
+        else:
+            anel = np.zeros(w.shape)
+        return coeff*np.sum(
+            np.square(dudx + dvdy + dwdz + anel))/2.0
+
+Since NumPy takes advantage of open source mathematics libraries that
```
Comment (ocefpaf) [inline papers/robert_jackson/paper.rst]: I'm not that is true. (I may be mistaken but as far as I know you would have to write a cython version or wrap that with numba to get parallelization, no?)

File: papers/robert_jackson/paper.rst [hunk 6]
```diff
@@ -0,0 +1,617 @@
...
+
+With one line of code, one can use the 3DVAR technique to retrieve winds using the
+:code:`pydda.retrieval.get_dd_wind_field` procedure.
+If one has a list of Py-ART grids :code:`list_of_grids` that they have loaded
+and provide :math:`\vec{\textbf{V}_{0}}` into arrays called
+:code:`u_init`, :code:`v_init`, and :code:`w_init`, retrieval of winds is as easy as
+
+.. code-block:: python
+
+    winds = pydda.retrieval.get_dd_wind_field(
```
Comment (ocefpaf) [inline papers/robert_jackson/paper.rst:212]: English is not my first language so forgive me if this comment does not make send but I have to say that the API where the words `retrieval.get_`  together is odd. I read that as get-get in my mind. Maybe the horse is out of the stable already would `compute/calculate` be better than `retrieval`?

File: papers/robert_jackson/paper.rst [hunk 7]
```diff
@@ -0,0 +1,617 @@
...
+    |                 | :code:`_era_interim`                         |
+    +-----------------+----------------------------------------------+
+    | Rawinsonde      | :code:`make_wind_field_from_profile`         |
+    +-----------------+----------------------------------------------+
+    | Constant field  | :code:`make_constant_wind_field`             |
+    +-----------------+----------------------------------------------+
+
+In addition, PyDDA includes 4 different initialization routines that will
+create this field for you from various data sources. In particular,
+PyDDA even supports the ECMWF web API for the automatic retrieval
```
Comment (ocefpaf) [inline papers/robert_jackson/paper.rst]: The word "retrieval" here is used in the sense of downloading, right?

File: papers/robert_jackson/paper.rst [hunk 8]
```diff
@@ -0,0 +1,617 @@
...
+
+The evaluation of :math:`J(\textbf{V})` can be done entirely using calls
+from NumPy and SciPy. For example, evaluating :math:`J_{c}(\vec{\textbf{V}}) =
+\nabla\cdot\vec{\textbf{V}})` with an optional anelastic term be reduced to a few NumPy calls:
+
+.. code-block:: python
+
+    import numpy as np
+
+    def calculate_mass_continuity(
```
Comment (ocefpaf) [inline papers/robert_jackson/paper.rst]: I like the code example but I have a question: does this take a lot of space in the final PDF? If so, should it be in an appendix? We have the equations and the code example is only to show how PyDDA does it.

File: papers/robert_jackson/paper.rst [hunk 9]
```diff
@@ -0,0 +1,617 @@
...
+    mygs[0] = const.make_constraint_from_era_interim(
+        my_grids)
+
+    # Make the output grids
+    ui, vi, wi = init.make_constant_wind_field(
+        grid_mhx, (0.0, 0.0, 0.0))
+    out_grids = pydda.retrieval.get_dd_wind_field_nested(
+        mygs, ui, vi, wi, Co=1.0, Cm=100.0,
+        Cmod=1e-5, model_fields=["hrrr", "erainterim"],
+        client=client)
```
Comment (ocefpaf) [inline papers/robert_jackson/paper.rst]: Suggestion: I would just mention that you can pass a `dask` `client` instance to PyDDA instead of the code snippet. (Or maybe putting the code snippet in an Appendix or an external link to it.)

File: papers/robert_jackson/paper.rst [hunk 10]
```diff
@@ -0,0 +1,617 @@
...
+:ref:`streamline` and :ref:`bighurricane` both show kilometer-scale
+regions of hurricane force winds that may otherwise not have been forecast
+to occur simply because they are outside of the primary region of damaging winds.
+This therefore shows the importance of having a high resolution, three dimensional
+wind retrieval when examining the effects of storm wind damage.
+
+Tornado in Sydney, Australia using 4 radars
+-------------------------------------------
+
+.. figure:: australian_radar_layout.png
```
Comment (ocefpaf) [inline papers/robert_jackson/paper.rst:457]: Suggestion: re-create the figure without cartopy's stock image. At this zoom level the pixels maybe be confused with relevant data, or use a [higher resolution version of the image](https://github.com/SciTools/cartopy/pull/815). (My first impression was: where is the colorbar, until I realized that was the stock-image.)

File: papers/robert_jackson/paper.rst [hunk 11]
```diff
@@ -0,0 +1,617 @@
...
+data from radar networks in order to retrieve the winds inside tornadoes.
+For example, a network of four scanning radars in the vicinity of Sydney,
+Australia captured a supercell within the vicinity of Sydney as shown in
+Figure :ref:`bomlayout`. In this retrieval, a horizontal domain of 350 km by
+550 km with 1 km grid spacing was used.
+
+.. figure:: Sydney_tornado.png
+    :align: center
+
+    A quiver plot inside a supercell that spawned a tornado in the vicinity of
```
Comment (ocefpaf) [inline papers/robert_jackson/paper.rst:474]: Can you increase the scale of the quiver plot arrows? They are barely visible.

**Review by ocefpaf — APPROVED** [review]
My final review is to approved the paper. All my comments are minor and/or clarification questions/suggestions.

The paper describes the development a python package for solving a problem in a domain within the scope of the SciPy conference. All the code present could be easily verified (I did not run all the examples though).

I did not verified the final PDF for length due to laptop limitations at the moment (travelling with a Windows machine). My review was based only on the `rst` file here.

File: papers/robert_jackson/paper.rst [hunk 12]
```diff
@@ -0,0 +1,617 @@
...
+* Support for integrating in data from the Rapid Refresh
+* Coarser resolution reanalyses such as the NCEP reanalysis as initializations and constraints.
+* Support for individual point analyses, such as those from wind profilers and METARs
+* Support for radar data in antenna coordinates
+* Improvements in visualizations
+* Documentation improvements, including better descriptions in the current English version of the documentation
+  and versions of the documentation in non-English languages.
+
+All contributions to PyDDA will have to be submitted by a pull request to the master branch
+on https://github.com/openradar/PyDDA. From there, the main developers will examine the pull
```
Comment (ocefpaf) [inline papers/robert_jackson/paper.rst:566]: Suggestion: I feel that the directions to the package should be in the first paragraph instead of buried in the "Contributor Information."

**Comment by nicholasmalaya** [comment]
I agree with @ocefpaf that this is a good submission to SciPy, and should be strongly considered for acceptance after clarifications and minor revisions. Thank you @rcjackson and co-authors!

A few board, major comments: 

- Are the examples provided? I don't see a link to them. 
- Should this PDF work in black and white? I'm of two minds. The pictures are pretty and viewing them on a computer isn't too unusual. In general as a reviewer I recommend papers work in black and white, however. 
- Similar to above, the plots aren't particularly friendly to the color-blind. Perhaps changing the contours to dot-dashed would help. 
- Is the variational problem regularized? Is this an option one should choose? 

I'll put more detailed comments in the .tex directly.

File: papers/robert_jackson/paper.rst [hunk 13]
```diff
@@ -0,0 +1,617 @@
...
+
+     J(\vec{\textbf{V}}) = J_{o}(\vec{\textbf{V}}) + J_{c}(\vec{\textbf{V}}) +
+                           J_{v}(\vec{\textbf{V}}) + J_{m}(\vec{\textbf{V}}) +
+                           J_{b}(\vec{\textbf{V}}) + J_{s}(\vec{\textbf{V}})
+
+where each addend is as in Table :ref:`costfunctions`.
+
+The evaluation of :math:`J(\textbf{V})` can be done entirely using calls
+from NumPy and SciPy. For example, evaluating :math:`J_{c}(\vec{\textbf{V}}) =
+\nabla\cdot\vec{\textbf{V}})` with an optional anelastic term be reduced to a few NumPy calls:
```
Comment (nicholasmalaya) [inline papers/robert_jackson/paper.rst]: Small typo here: missing left paranthesis. I recommend:
```suggestion
(\nabla\cdot\vec{\textbf{V}})` with an optional anelastic term be reduced to a few NumPy calls:
```

File: papers/robert_jackson/paper.rst [hunk 14]
```diff
@@ -0,0 +1,617 @@
...
+reanalysis to provide the most complete wind retrieval possible.
+
+Contributor Information
+-----------------------
+
+We are currently welcoming contributions from the community into PyDDA. A PyDDA road map
+demonstrates what kinds of contributions to PyDDA would be useful. As of the writing
+of this paper, the road map states that the current goals of PyDDA are to implement:
+
+* Support for a greater number of high resolution (LES) models such as CM1
```
Comment (nicholasmalaya) [inline papers/robert_jackson/paper.rst]: Citation for CM1?

File: papers/robert_jackson/paper.rst [hunk 15]
```diff
@@ -0,0 +1,617 @@
...
+PyDDA even supports the ECMWF web API for the automatic retrieval
+of ERA-Interim reanalysis data. These various routines are listed in the
+Table :ref:`inits`.
+
+.. table:: The differing model constraints PyDDA can provide to the user. :label:`consts`
+
+    +------------------+----------------------------------------------+
+    | Data source      | Routine in constraints module                |
+    +==================+==============================================+
+    | Weather Research | :code:`make_constraint_from_wrf`             |
```
Comment (nicholasmalaya) [inline papers/robert_jackson/paper.rst:267]: I'd prefer citations for these model constraints, or defining what they are in the text. This is important for reproducability, e.g. stating a model or constraint is seldom sufficient to define it for a user. I'd also recommend a single sentence for each be added to the document (or, better, the table label) that notes roughly what these model constraints are.

File: papers/robert_jackson/paper.rst [hunk 16]
```diff
@@ -0,0 +1,617 @@
...
+Since :code:`pydda.retrieval.get_dd_wind_field` has 3D NumPy arrays as inputs
+for the initialization, this allows the user to enter in an arbitrary NumPy
+array with the same shape as the analysis grid as the initialization field.
+
+.. table:: The differing initalizations PyDDA can provide to the user. :label:`inits`
+
+    +-----------------+----------------------------------------------+
+    | Data source     | Routine in initialization module             |
+    +=================+==============================================+
+    | Weather         |                                              |
```
Comment (nicholasmalaya) [inline papers/robert_jackson/paper.rst:238]: I'd prefer citations for these initializations. I'd also recommend a single sentence for each be added to the document (or, better, the table label) that notes roughly what these initializations are.

File: papers/robert_jackson/paper.rst [hunk 17]
```diff
@@ -0,0 +1,617 @@
...
+
+Figure :ref:`somanyradars` shows the resulting wind field
+of such a retrieval during a case of stratiform rain with embedded convection
+that occurred over the SGP site on 04 October 2017. Generally,
+weaker winds and a less organized structure is seen compared to the
+previous two examples. This would be expected in such conditions.
+However, this also demonstrates the success in integrating radar data
+from 3 radars and a high resolution
+reanalysis to provide the most complete wind retrieval possible.
+
```
Comment (nicholasmalaya) [inline papers/robert_jackson/paper.rst:548]: I recommend you add a new section here. It would be 'verification' and discuss what regression testing, correctness checks, etc. exist in PyDDA. After all, how can users know if this library is functioning correctly? This is important for reproducibility of the work.

**Comment by cchoirat** [comment]
I mostly have minor comments about the paper (from the PDF version I retrieved initially), which would be a good submission for SciPy.

- There are repetitions, on p.1., for example, "to use as it uses" or "In addition, the addition"

- Sentences such as "This therefore permits the easy installation of PyDDA using pip of anaconda.  Given that alone is a major hurdle to [...]", I think, read the opposite of of they actually mean.

- p. 4, "Figure using a similar" : Figure number is missing.

File: papers/robert_jackson/paper.rst (hunk 1 above)
Comment (rcjackson) [reply papers/robert_jackson/paper.rst:30]: In the meteorology community, a “retrieval” is defined as calculating a quantity from a set of other quantities that are recorded by the instrument. For example, if you want to retrieve sea surface temperature from satellite, this is not detected directly, but rather derived from irradiances from various channels scanned from the satellite. In this case, we are deriving three dimensional winds from the radial velocities detected by the radar. This therefore explains our choice of wording here.

File: papers/robert_jackson/paper.rst (hunk 2 above)
Comment (rcjackson) [reply papers/robert_jackson/paper.rst]: We have changed this to say “non-thread-safe” as that is the actual spelling of the word.

Comment (rcjackson) [reply papers/robert_jackson/paper.rst]: We intended this to be understandable by a general scientific audience. Therefore, we took care to explain any necessary jargon that may not be known outside of the radar meteorology and optimization communities.

File: papers/robert_jackson/paper.rst (hunk 4 above)
Comment (rcjackson) [reply papers/robert_jackson/paper.rst]: We have changed the phrasing of this sentence to the reviewer’s suggestion.

File: papers/robert_jackson/paper.rst (hunk 2 above)
Comment (rcjackson) [reply papers/robert_jackson/paper.rst]: We have changed the phrasing of this sentence to the reviewer’s suggestion.

File: papers/robert_jackson/paper.rst (hunk 7 above)
Comment (rcjackson) [reply papers/robert_jackson/paper.rst]: We have changed the second “in addition” to “moreover”.

File: papers/robert_jackson/paper.rst (hunk 8 above)
Comment (rcjackson) [reply papers/robert_jackson/paper.rst]: This code takes almost half of a page of the PDF, so we decided to move this code to the appendix so for the interested reader.

File: papers/robert_jackson/paper.rst (hunk 10 above)
Comment (rcjackson) [reply papers/robert_jackson/paper.rst:457]: We have removed the cartopy stock figure from the image.

File: papers/robert_jackson/paper.rst (hunk 11 above)
Comment (rcjackson) [reply papers/robert_jackson/paper.rst:474]: The size of the quiver arrows has been increased.

File: papers/robert_jackson/paper.rst (hunk 12 above)
Comment (rcjackson) [reply papers/robert_jackson/paper.rst:566]: We have added a URL to the package to the abstract and the introduction.

File: papers/robert_jackson/paper.rst (hunk 9 above)
Comment (rcjackson) [reply papers/robert_jackson/paper.rst]: We have removed the code snippet in this section and simply mention that a dask Client can be passed into PyDDA for the nested retrieval.

File: papers/robert_jackson/paper.rst (hunk 3 above)
Comment (rcjackson) [reply papers/robert_jackson/paper.rst:48]: Our goal is to make this work accessible to the general scientific community, so we avoided the use of or explained jargon as it was needed.

File: papers/robert_jackson/paper.rst (hunk 6 above)
Comment (rcjackson) [reply papers/robert_jackson/paper.rst]: f you use versions of NumPy that are built against libraries such as OpenBLAS, or Intel’s Math Kernel Library, then NumPy will take advantage of parallelization because it will use the parallel capabilities in these libraries to perform the calculations. For more information, see this link:

https://markus-beuckelmann.de/blog/boosting-numpy-blas.html

However, since not all configurations of numpy use these libraries, we have changed the phrasing here to “Since NumPy can be configured to take advantage of open source mathematics libraries that parallelize the calculation.”

Comment (rcjackson) [reply papers/robert_jackson/paper.rst:212]: At this point changing the name of the retrieval methods will require a new release which while is planned to happen right before SciPy, that may not be done in time for submission of the conference proceedings. I have added an issue to PyDDA’s GitHub page regarding this. Such a fix is definitely within PyDDA’s roadmap, in particular with regards to having the program and documentation understandable to those people where English is not their first language, so we thank the reviewer for his valuable comment.

File: papers/robert_jackson/paper.rst (hunk 7 above)
Comment (rcjackson) [reply papers/robert_jackson/paper.rst]: We have retaiored this whole section to focus more on the steps used to create the initializations and constraints in PyDDA at the suggestion of @nicholasmalaya. In doing this, this sentence was eliminated.

File: papers/robert_jackson/paper.rst (hunk 13 above)
Comment (rcjackson) [reply papers/robert_jackson/paper.rst]: We have fixed this parenthesis.

File: papers/robert_jackson/paper.rst (hunk 14 above)
Comment (rcjackson) [reply papers/robert_jackson/paper.rst]: We have added a citation to Bryan and Fritsch 2002 which is the original paper for CM1:

Bryan, G.H. and J.M. Fritsch, 2002: A Benchmark Simulation for Moist Nonhydrostatic Numerical Models. Mon. Wea. Rev., 130, 2917–2928, https://doi.org/10.1175/1520-0493(2002)130<2917:ABSFMN>2.0.CO;2

File: papers/robert_jackson/paper.rst (hunk 15 above)
Comment (rcjackson) [reply papers/robert_jackson/paper.rst:267]: Inserting models as constraints into variational wind retrievals is a new feature of PyDDA and is something that has not been commonly done in multiple-Doppler retrievals. Therefore, we have decided to provide further details about how both the initializations and constraints are constructed. The details provided include how the data are interpolated to the analysis grid and then how the extra term related to the model constraint in the cost function is constructed when performing the optimization.

File: papers/robert_jackson/paper.rst (hunk 16 above)
Comment (rcjackson) [reply papers/robert_jackson/paper.rst:238]: Similar to the previous comment we have added more details about how the initializations from model data are constructed. In the table caption we have added:

“These constraints are constructed by interpolating the model :math:`J(\vec{\textbf{V}})` to the analysis grid coordinates.”

In addition, in the following paragraph we have stated how the model initialization is constructed by interpolating the three components of the wind field from the model onto the analysis grid using nearest-neighbor interpolation and then using that as the initial guess for the optimization loop.

File: papers/robert_jackson/paper.rst (hunk 17 above)
Comment (rcjackson) [reply papers/robert_jackson/paper.rst:548]: We run a suite of 27 tests on PyDDA using pytest. Every time a pull request to the master branch of PyDDA is requested, Travis CI will run this suite of tests to ensure that PyDDA is functioning correctly. We have added a section on verification that both states that we run this suite of tests and gives a few examples of the unit tests that are performed on PyDD. Examples of such unit tests check to see if the mass continuity cost function is nonzero in a converging wind field, the smoothness cost function is nonzero in a discontinuous wind field, and if the variational technique converges to the model input if only the model field and constraint are used. This helps to ensure that the variational technique is working as intended.

**Comment by rcjackson** [comment]
I am going to place the general response to the 3 reviewers here, with responses to their more specific comments in the relevant comment sections in the thread above. 

@ocepaf:
We thank the reviewer for taking the time to provide his valuable feedback. We have responded to the individual comments using GitHub’s commenting system. 

@nicholasmalaya:
 We thank @nicholasmalaya for his valuable comments. They have considerably improved the manuscript.
We have included the source code and the links to the data in all of the examples in this document here.
https://openradarscience.org/PyDDA/source/auto_examples/index.html. 

The figures have been regenerated to be more readable in response to all of the reviewers’ comments. In particular, the contours in the plots have been switched to hatches. In addition, we now use @CameronHomeyer’s colormap that visualizes radar reflectivity fields using a color-blind friendly scheme which will make the figures more accessible.  This colormap is now used in Figures 2, 3, 4, 6, and 8. In addition, the sizes of the quivers and streamline arrows have been enhanced in the new versions of Figure 1 and 6, making these figures more readable. Finally, Figure 6 now uses a hatched contour instead of a transparent contour to denote the regions where updrafts are present.

In terms of trying to make this paper readable in black and white, it is no longer usual in the meteorology community to attempt to make radar images work in black and white as they are very difficult to visualize effectively without color. In essence, any colormap we use has to highlight regions of light rain, moderate rain, heavy rain, and hail all in ways that are easily distinguishable. This is very difficult to do effectively in black and white. Therefore, in order to better highlight significant features of the storm, we have decided that color figures are the best way to visualize the data. The use of color-blind friendly colormaps does help the figures appear better in black and white.

The variational problem is regularized in two different ways in order to ensure a physically realistic solution that is free of noise. One is through the use of a hard box constraint on the solution when executing the L-BFGS-B minimization technique. Throughout the domain, the wind velocities are constrained to be within -100 to 100 m/s which is a physically realistic scenario for virtually all weather that occurs on Earth. In addition, in order to reduce noise in the final solution, there are options to have a low pass filter as well as adjusting the smoothness constraint that is directly proportional to the Laplacian of the wind field. We have added a few sentences in the paper that mention these techniques.

@cchoirat:

We thank the reviewer for taking the time to review the paper and providing her valuable comments. We have embedded our responses below.

There are repetitions, on p.1., for example, "to use as it uses" or "In addition, the addition"
The repetitiveness of the introduction was duly noted by all reviewers so some changes here were made to improve the readability of this section For example, we have changed the reviewer’s first example of repetition to state “While CEDRIC was revolutionary for its time, it is difficult to use as a separate scripting language is the input for the retrieval.” In addition, the first “in addition” was replaced with “moreover” following the suggestion of @ocepaf. 


Sentences such as "This therefore permits the easy installation of PyDDA using pip of anaconda. Given that alone is a major hurdle to [...]", I think, read the opposite of of they actually mean.

To make this clearer we have replaced “this” with “installation” in this sentence.

p. 4, "Figure using a similar" : Figure number is missing.

We have fixed the figure number here.

File: papers/robert_jackson/paper.rst (hunk 6 above)
Comment (ocefpaf) [reply papers/robert_jackson/paper.rst:212]: Looks like I cannot resolve this comment and it will be hard to follow what was addressed and what needs work. @deniederhut is there a way to grant reviewers the power to "resolve" conversations?

File: papers/robert_jackson/paper.rst (hunk 1 above)
Comment (ocefpaf) [reply papers/robert_jackson/paper.rst:30]: I guess that as an oceanographer I prefer calculate rather than retrieve :smile:

File: papers/robert_jackson/paper.rst (hunk 6 above)
Comment (rcjackson) [reply papers/robert_jackson/paper.rst:212]: I can resolve conversations (which makes no sense, I think it should be the editor who does it). Do you want me to resolve it?

Comment (ocefpaf) [reply papers/robert_jackson/paper.rst:212]: To make it easier for a second pass/conclusion please "resolve" all of my comments where I added the :+1: to your answer.

File: papers/robert_jackson/paper.rst (hunk 15 above)
Comment (nicholasmalaya) [reply papers/robert_jackson/paper.rst:267]: This is a valuable addition to the paper and materially improves the reproducability and traceability of the work. Thank you for adding this! I know it was a non-trivial request.

File: papers/robert_jackson/paper.rst (hunk 17 above)
Comment (nicholasmalaya) [reply papers/robert_jackson/paper.rst:548]: This is very good and a useful addition to the paper. I have one minor quibble: the section is now labelled 'Validation' but the section is actually more broad. For example, you note '27 unit tests' (verification). I would change the section title to 'Verification and Validation' or something sufficiently broad, such as 'software correctness testing', etc. 

Thank you again for adding this portion.

**Comment by nicholasmalaya** [comment]
@deniederhut My final review of this paper is 'strong accept'. The authors have quickly addressed all my comments, some of which were non-trivial to produce, of which I am appreciative. The paper is improved and conforms to many of the SciPy guidelines for reproducability and openness. This is a useful contribution and should be of interest to the SciPy community. 

@rcjackson thank you for the edits to the colormaps, which has helped the paper. I understand and accept your comments about the inability to use greyscale in meterology. Thank you for the additions to the variational problem. Thank you and your co-authors for your submission.

**Comment by ocefpaf** [comment]
> @deniederhut My final review of this paper is 'strong accept'. The authors have quickly addressed all my comments, some of which were non-trivial to produce, of which I am appreciative. The paper is improved and conforms to many of the SciPy guidelines for reproducability and openness. This is a useful contribution and should be of interest to the SciPy community.

Same here. Strong accept.

**Comment by cchoirat** [comment]
Same for me.

**Comment by deniederhut** [comment]
Awesome. Thanks everyone!

**Comment by rcjackson** [comment]
@deniederhut  Scott Collis, Timothy Lang, and I all work for government agencies that do not allow authors to retain copyrights on publications but must transfer it over to our institution. How

Also, NASA is wanting a copy with the "DRAFT" watermark removed for their final approval, and apparently are wanting it now. How would I go about removing the watermark and change the copyright notice for purposes of the approval?

**Comment by deniederhut** [comment]
Hm. So the copyright is not really up to me to change. It's CC-BY anyway, so while you technically hold the copyright, the only protection that gives you is the right to be cited. One possible solution would be to list NASA explicitly as the first author (the Jupyter team does something similar), but the copyright would be to NASA et al. -- @sjsrey, do you have thoughts about this? 

I'll add for context that we regularly receive submissions from government agencies (including NASA, NOAA, NWS, NGS, LANL, SNL, the army, etc.) and it hasn't been a concern for at least the three years I've been on this committee.

As for the watermark, we swap in different latex styles for draft, conference-ready, and final. If you need it now now, you can build your paper with `publisher/static/ready.sty`. If you can wait until next week, I can generate it when I do all the other conference-ready papers.

**Comment by scollis** [comment]
@rcjackson I think we may be good here.. SciPy does not want to claim copyright so, in panda, the correct classification is
```No copyright is being asserted and there is no license to be completed.```

This essentially means copyright is retained by the institutions and individuals. 
I think  @tjlang's issues can just be fixed with a tweak to the tex template

**Comment by rcjackson** [comment]
I know we're good on ANL's side. It seems to be that NASA more has an issue with the copyright statement being there at all. Would it be good to do something like "copyright authors?" I'm contacting the approval folks over at NASA to see if this would suffice. NASA has started to become more strict with every submission in the past year or so, which is the reason we're having this issue today.

**Comment by rcjackson** [comment]
I just talked to their approval folks and they said it is probably ok to leave the copyright statement as it is after I explained it to them that we are not transferring copyright. So go ahead and build as normal until I hear otherwise. I sent them the copy with the watermark removed.
